wasm/wasm_source/src/vp_user.rs

```diff
@@ -628,6 +629,153 @@ mod tests {
+            key::testing::keypair_3().ref_to(),
+            key::testing::keypair_4().ref_to(),
+        );
+        let genesis_validators = [
+            GenesisValidator {
+                address: validator1.clone(),
+                tokens: stake1,
+                consensus_key: ck1.clone(),
+                protocol_key: protocol_key.clone(),
+                commission_rate,
+                max_commission_rate_change,
+                eth_hot_key: eth_hot_key.clone(),
+                eth_cold_key: eth_cold_key.clone(),
+                metadata: Default::default(),
+            },
+            GenesisValidator {
+                address: validator3.clone(),
+                tokens: stake3,
+                consensus_key: ck3.clone(),
+                protocol_key: protocol_key.clone(),
+                commission_rate,
+                max_commission_rate_change,
+                eth_hot_key: eth_hot_key.clone(),
+                eth_cold_key: eth_cold_key.clone(),
+                metadata: Default::default(),
+            },
+            GenesisValidator {
+                address: validator2.clone(),
+                tokens: stake2,
+                consensus_key: ck2,
+                protocol_key,
+                commission_rate,
+                max_commission_rate_change,
+                eth_hot_key,
+                eth_cold_key,
+                metadata: Default::default(),
+            },
+        ];
+
+        println!("\nValidator1: {}", &validator1);
+        println!("Validator2: {}", &validator2);
+        println!("Validator3: {}\n", &validator3);
+
+        // Init PoS storage
+        init_pos(&genesis_validators[..], &pos_params, Epoch(0));
+
+        // Initialize a tx environment
+        let mut tx_env = tx_host_env::take();
+        let token = address::testing::nam();
+
+        // write the denomination of NAM into storage
+        token::write_denom(
+            &mut tx_env.state,
+            &token,
+            token::NATIVE_MAX_DECIMAL_PLACES.into(),
+        )
+        .unwrap();
+
+        // Jail validator3
+        jail_validator(
+            &mut tx_env.state,
+            &pos_params,
+            &validator3,
+            Epoch(0),
+            Epoch(0),
+        )
+        .unwrap();
+
+        // Initialize VP environment
+        vp_host_env::init_from_tx(validator3.clone(), tx_env, |_address| {
+            // Unjail validator3
+            tx::ctx().unjail_validator(&validator3).unwrap()
+        });
+
+        let pks_map = AccountPublicKeysMap::from_iter(vec![ck3]);
+
+        let mut vp_env = vp_host_env::take();
+        let mut tx_data = Tx::from_type(TxType::Raw);
+        tx_data.set_data(Data::new(vec![]));
+        tx_data.set_code(Code::new(vec![], None));
+        tx_data.add_section(Section::Signature(Signature::new(
+            vec![tx_data.raw_header_hash()],
+            pks_map.index_secret_keys(vec![key::testing::keypair_4()]),
```

Comment:
just a little nit about this test - it would make it clearer to use the `ck3` for this directly to see that it's being signed by `validator3` (the `ck`s are converted to PKs with `ref_to` so we should keep them as SKs and turn them into PKs as needed)

Reply:
done in fc4e7488183c5f4718236cf5a682ba38889b4b5b